src/QtHtmlReader.cpp

```diff
@@ -0,0 +1,250 @@
+/**
+ * @file
+ * @brief Source file for QtHtmlReader class
+ * @author Jonathan Thomas <jonathan@openshot.org>
+ *
+ * @ref License
+ */
+
+/* LICENSE
+ *
+ * Copyright (c) 2008-2019 OpenShot Studios, LLC
+ * <http://www.openshotstudios.com/>. This file is part of
+ * OpenShot Library (libopenshot), an open-source project dedicated to
+ * delivering high quality video editing and animation solutions to the
+ * world. For more information visit <http://www.openshot.org/>.
+ *
+ * OpenShot Library (libopenshot) is free software: you can redistribute it
+ * and/or modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * OpenShot Library (libopenshot) is distributed in the hope that it will be
+ * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with OpenShot Library. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "../include/QtHtmlReader.h"
+#include <QImage>
+#include <QPainter>
+#include <QTextDocument>
+#include <QGuiApplication>
+#include <QAbstractTextDocumentLayout>
+
+using namespace openshot;
+
+/// Default constructor (blank text)
+QtHtmlReader::QtHtmlReader() : width(1024), height(768), x_offset(0), y_offset(0), html(""), background_color("#000000"), is_open(false), gravity(GRAVITY_CENTER)
+{
+	// Open and Close the reader, to populate it's attributes (such as height, width, etc...)
+	Open();
+	Close();
+}
+
+QtHtmlReader::QtHtmlReader(int width, int height, int x_offset, int y_offset, GravityType gravity, std::string html, std::string background_color)
+: width(width), height(height), x_offset(x_offset), y_offset(y_offset), gravity(gravity), html(html), background_color(background_color), is_open(false)
+{
+	// Open and Close the reader, to populate it's attributes (such as height, width, etc...)
+	Open();
+	Close();
+}
+
+// Open reader
+void QtHtmlReader::Open()
+{
+	// Open reader if not already open
+	if (!is_open)
+	{
+		// create image
+		image = std::shared_ptr<QImage>(new QImage(width, height, QImage::Format_RGBA8888));
+		image->fill(QColor(background_color.c_str()));
+
+		//start painting
+		QPainter painter;
+		if (!painter.begin(image.get())) {
+			return;
+		}
+
+		//set background
+		painter.setBackground(QBrush(background_color.c_str()));
+
+		//draw text
+		QTextDocument text_document;
+		text_document.setTextWidth(width);
+		text_document.setHtml(html.c_str());
```

Comment:
Oh yeah, I knew there was something I forgot. [Qt's docs say](https://doc.qt.io/qt-5/qtextdocument.html#undoRedoEnabled-prop) that QTextDocument has an undo/redo stack, and it's _enabled_ by default? 

I assume there must be some overhead associated with that. (Though it doesn't seem to be very well-documented just _how much_ it might be — I'm also not clear on whether actions are automatically added to the stack, or only when explicitly added by the code.) Since the caller has no access to the QTextDocument, and **we** don't use the stack, a `text_document.setUndoRedoEnabled(false);` is probably worth doing here.

Reply:
Done